### PR TITLE
Sample and Guide for arcgis.learn TextClassifier module

### PR DESCRIPTION
**Pull request by anurag89sharma**
Added Guide & Sample notebook for `arcgis.learn`'s `TextClassifier` module

### CODE REVIEW

**Comment by priyankatuteja** [comment]
@akhilnegi Please do the first round of reviews

**Comment by priyankatuteja** [comment]
@anurag89sharma For the sample notebook, the csv can be zipped and published as an item on api data owner account. Please refer to Madison crime sample notebook to see how data is being provided to ensure that these notebooks can be run on try it live.

File: guide/14-deep-learning/text-classification-workflow.ipynb [hunk 1]
```diff
@@ -0,0 +1,1239 @@
+{
```
Comment (akhilnegi) [inline guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>The TOC can be restructured as per below suggestions:</p><p> </p><ul><li>The model uses transformer architecture can be emphasized in <a href="https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/files/#Introduction" target="_blank">Introduction</a> and <a href="https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/files/#Transformer-Basics" target="_blank">Transformer Basics</a> can be made a top level heading.</li><li><a href="https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/files/#Prerequisites" target="_blank">Prerequisites</a> can be the software and data requirements.  </li><li><a href="https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/files/#How-to-choose-an-appropriate-model-for-your-dataset?" target="_blank">How to choose an appropriate model for your dataset?</a> can be a sub under TextClassifier model heading.</li><li><a href="https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/files/#Supported-formats-for-labeled-training-data" target="_blank">Supported formats for labeled training data</a> can be moved under data preparation.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (anurag89sharma) [reply guide/14-deep-learning/text-classification-workflow.ipynb:1]: Done the necessary changes,

Comment (akhilnegi) [reply guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>Few typo corrections on the sample NB suggested in the below doc:</p><p> </p><p>https://esriis.sharepoint.com/:w:/s/EsriRDCenterNewDelhi/EeebPFZsfPhCs2yOcB6CiKUBzxb-LdNdaH-4h0PjUKb0Lw?e=LsmbF6</p>

File: samples/04_gis_analysts_data_scientists/identifying-county-names-from-incomplete-house-addresses.ipynb [hunk 2]
```diff
@@ -0,0 +1,1903 @@
+{
```
Comment (akhilnegi) [inline samples/04_gis_analysts_data_scientists/identifying-county-names-from-incomplete-house-addresses.ipynb:1]: <p>Seems to be an issue with the output.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (akhilnegi) [inline samples/04_gis_analysts_data_scientists/identifying-county-names-from-incomplete-house-addresses.ipynb:1]: <p>Description talks about an error rate, but that is not seen in the output.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

Comment (akhilnegi) [inline samples/04_gis_analysts_data_scientists/identifying-county-names-from-incomplete-house-addresses.ipynb:1]: <p>This ("In this notebook we will build a classifier using <code>TextClassifier</code> class of <code>arcgis.learn.text</code> module.") needs to be in past tense.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

File: guide/14-deep-learning/text-classification-workflow.ipynb (hunk 1 above)
Comment (anurag89sharma) [reply guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>Done the necessary changes</p>

File: samples/04_gis_analysts_data_scientists/identifying-county-names-from-incomplete-house-addresses.ipynb (hunk 2 above)
Comment (anurag89sharma) [reply samples/04_gis_analysts_data_scientists/identifying-county-names-from-incomplete-house-addresses.ipynb:1]: <p>Seems like a rendering issue on app.reviewnb.com. Actual notebook doesn't show <code>&lt;IPython.core.display.HTML object&gt;</code> this line</p>

Comment (anurag89sharma) [reply samples/04_gis_analysts_data_scientists/identifying-county-names-from-incomplete-house-addresses.ipynb:1]: fixed the description

Comment (anurag89sharma) [reply samples/04_gis_analysts_data_scientists/identifying-county-names-from-incomplete-house-addresses.ipynb:1]: fixed

**Review by akhilnegi — COMMENTED** [review]
Looks good to me.

File: guide/14-deep-learning/text-classification-workflow.ipynb [hunk 3]
```diff
@@ -0,0 +1,1222 @@
...
```
Comment (akhilnegi) [inline guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>In the first bullet it entity extraction to be replaced with text classification.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (anurag89sharma) [reply guide/14-deep-learning/text-classification-workflow.ipynb:1]: fixed

Comment (AtmaMani) [inline guide/14-deep-learning/text-classification-workflow.ipynb:1]: <ol><li>Text classification also known as text tagging or text categorization is the process of assigning tags/labels to <s>an</s> unstructured text.</li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (AtmaMani) [inline guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>"Transformers in NLP are novel architectures that aims to solve sequence-to-sequence tasks while handling long-range dependencies with ease." -&gt; I think this line can either be elaborated on what 'sequence-to-sequence' tasks are and what long range deps are. If that is too complex within the confines of a single notebook, please hyperlink to a reference.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (AtmaMani) [inline guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>For the multi-text class image, can you update it such that the zeros and ones are randomly spread? It is not clear to me if the class values are binary (0 /1 ) or if they can take numeric values (greater than 1, perhaps corresponding to the number of times each class occurs in the text). May be you can explain this in the text following the image.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (AtmaMani) [inline guide/14-deep-learning/text-classification-workflow.ipynb:1]: <ol><li><a href="https://huggingface.co/transformers/v3.0.2/pretrained_models.html" target="_blank">This page</a> mentions different <strong>transformer<s>s</s> </strong>architectures <a href="https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/files/#References" target="_blank">[4]</a></li><li>The hyperlinks for references need to be relative, instead they point to absolute links on nbviewer</li><li>If your dataset consists of text in multiple languages you can <strong>choose<s>s</s></strong> models mentioned in the above link.</li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

File: samples/04_gis_analysts_data_scientists/identifying-country-names-from-incomplete-house-addresses.ipynb [hunk 4]
```diff
@@ -0,0 +1,1903 @@
+{
```
Comment (AtmaMani) [inline samples/04_gis_analysts_data_scientists/identifying-country-names-from-incomplete-house-addresses.ipynb:1]: <p>The data licenses can be found in <code>data/country-classifier/LICENSE.txt</code>. -&gt; Replace saying the data will be downloaded from an ArcGIS Item.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/798/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

**Comment by AtmaMani** [comment]
@anurag89sharma I think the file ' ' can be renamed to 'Text classification with arcgis.learn' (similar to [object detection](https://developers.arcgis.com/python/guide/object-detection/)). Could you change both the file names and title? You may have to fix the hyperlink in the sample notebook after this.

File: guide/14-deep-learning/text-classification-workflow.ipynb (hunk 3 above)
Comment (anurag89sharma) [reply guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>fixed</p>

Comment (anurag89sharma) [reply guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>Added appropriate hyperlinks to <strong>sequence-to-sequence</strong> and <strong>long-range dependencies</strong> terms</p>

Comment (anurag89sharma) [reply guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>The dataset from where the image is taken is from <a href="https://www.kaggle.com/c/jigsaw-toxic-comment-classification-challenge" target="_blank">Toxic comment classification challenge</a>, which contains comments with very strong and sometimes abusive words. We have tried not to show such text which taking the image as the content would not be appropriate for our users. May be we can keep the image as it is. I agree with your 2nd point. I will be making necessary changes to the text following the image</p>

File: samples/04_gis_analysts_data_scientists/identifying-country-names-from-incomplete-house-addresses.ipynb (hunk 4 above)
Comment (anurag89sharma) [reply samples/04_gis_analysts_data_scientists/identifying-country-names-from-incomplete-house-addresses.ipynb:1]: <p>For the dataset we have picked up for the sample, mentioning the data license is required. What would be the best way to do so here?</p>

File: guide/14-deep-learning/text-classification-workflow.ipynb (hunk 3 above)
Comment (anurag89sharma) [reply guide/14-deep-learning/text-classification-workflow.ipynb:1]: <p>fixed point 1 &amp; 3. For point 2, not sure why the hyper links are not working on nbviewer. I used the same code/procedure to create the hyper link as in the <a href="https://developers.arcgis.com/python/guide/how-named-entity-recognition-works/#Supported-formats-for-labeled-training-data" target="_blank">guide for entity recognition</a>. One can see how clicking to the <strong>Docanno [2]</strong> (last point) link will send the user the desired section on the web page. </p>

File: samples/04_gis_analysts_data_scientists/identifying-country-names-from-incomplete-house-addresses.ipynb (hunk 4 above)
Comment (AtmaMani) [reply samples/04_gis_analysts_data_scientists/identifying-country-names-from-incomplete-house-addresses.ipynb:1]: <p>I get it now, its the relative location within the data archive. I misunderstood it earlier</p><p> </p>

**Review by AtmaMani — APPROVED** [review]
